From 9ac0ee140ce13e9594811dfa8217db0b150e33d9 Mon Sep 17 00:00:00 2001
From: Julien STEPHAN <jstephan@baylibre.com>
Date: Fri, 5 Mar 2021 09:28:40 +0000
Subject: [PATCH] add external nnapi delegate

Add an external nnapi delegate to wrap the nnapi delegate, so the nnapi
delegate can be used with python bindings

Signed-off-by: Julien STEPHAN <jstephan@baylibre.com>
---
 .../utils/nnapi_external_delegate/BUILD       | 27 ++++++++++++++
 .../external_delegate_adaptor.cc              | 37 +++++++++++++++++++
 2 files changed, 64 insertions(+)
 create mode 100644 tensorflow/lite/delegates/utils/nnapi_external_delegate/BUILD
 create mode 100644 tensorflow/lite/delegates/utils/nnapi_external_delegate/external_delegate_adaptor.cc

diff --git a/tensorflow/lite/delegates/utils/nnapi_external_delegate/BUILD b/tensorflow/lite/delegates/utils/nnapi_external_delegate/BUILD
new file mode 100644
index 00000000000..ac2a7dbfb39
--- /dev/null
+++ b/tensorflow/lite/delegates/utils/nnapi_external_delegate/BUILD
@@ -0,0 +1,27 @@
+load("//tensorflow/lite:build_def.bzl", "tflite_copts")
+load("//tensorflow/lite/tools/evaluation/tasks:build_def.bzl", "task_linkopts")
+
+package(
+    default_visibility = [
+        "//visibility:public",
+    ],
+    licenses = ["notice"],  # Apache 2.0
+)
+
+cc_binary(
+    name = "nnapi_external_delegate.so",
+    srcs = [
+        "external_delegate_adaptor.cc",
+    ],
+    linkopts = [
+        "-Wl,-soname=nnapi_external_delegate.so",
+    ],
+    linkshared = 1,
+    linkstatic = 1,
+    deps = [
+        "//tensorflow/lite/delegates/nnapi:nnapi_delegate",
+        "//tensorflow/lite/c:common",
+        "//tensorflow/lite/tools:command_line_flags",
+        "//tensorflow/lite/tools:logging",
+    ],
+)
diff --git a/tensorflow/lite/delegates/utils/nnapi_external_delegate/external_delegate_adaptor.cc b/tensorflow/lite/delegates/utils/nnapi_external_delegate/external_delegate_adaptor.cc
new file mode 100644
index 00000000000..dcc2708999b
--- /dev/null
+++ b/tensorflow/lite/delegates/utils/nnapi_external_delegate/external_delegate_adaptor.cc
@@ -0,0 +1,37 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/delegates/nnapi/nnapi_delegate.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Defines two symbols that need to be exported to use the TFLite external
+// delegate. See tensorflow/lite/delegates/external for details.
+TFL_CAPI_EXPORT TfLiteDelegate* tflite_plugin_create_delegate(
+    char** options_keys, char** options_values, size_t num_options,
+    void (*report_error)(const char*)) {
+  static tflite::StatefulNnApiDelegate* delegate = new tflite::StatefulNnApiDelegate();
+  return delegate;
+}
+
+TFL_CAPI_EXPORT void tflite_plugin_destroy_delegate(TfLiteDelegate* delegate) {
+	delete reinterpret_cast<tflite::StatefulNnApiDelegate*>(delegate);
+}
+
+#ifdef __cplusplus
+}
+#endif  // __cplusplus
-- 
2.17.1

